### PR TITLE
refactor: linkcheck: Separate worker feature from builder class

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Refactoring

### Purpose
- Now, the linkcheck builder becomes bigger and complicated.
- To reduce the complexity of the class, this tries to separate the worker feature from the builder class as a first step.
- I'll work to separate the thread-conductor feature from the builder on the next step.
- Finally, the builder will be changed to the representer feature (logging and output).

### CODE REVIEW

File: sphinx/builders/linkcheck.py
Comment (tk0miya) [inline sphinx/builders/linkcheck.py:184]: Note: The methods below are not changed at all. I only moved (copied) `check_thread()` and `limit_rate()` to `HyperlinkAvailabilityCheckWorker` class.

File: sphinx/builders/linkcheck.py
Comment (tk0miya) [inline sphinx/builders/linkcheck.py:288]: I'll refactor these attributes if agreed.

File: sphinx/builders/linkcheck.py
Comment (francoisfreitag) [inline sphinx/builders/linkcheck.py]: <s>This is not compatible with extensions, as `good` no longer collects the good links. Similarly, `broken` and `redirects` won’t be populated.

Should the change target the master branch?</s>

**Review by francoisfreitag — COMMENTED** [review]
:+1: I like the idea of splitting the responsibilities, the builder currently does many things.

> I'll work to separate the thread-conductor feature from the builder on the next step.

Sounds like a good plan to me, especially for state shared between workers (rate limits) but not tied to the builder.

File: sphinx/builders/linkcheck.py
Comment (francoisfreitag) [reply sphinx/builders/linkcheck.py]: I missed commit refactor: “linkcheck: Do not cache the checking result”, which explains this change in behavior is expected.

---

> This completely deprecates `broken`, `good` and `redirected`, but allows to access empty ones via properties not to break 3rd party extensions.

I wonder if that’s really better than removing the properties? The empty properties are likely to give incorrect results to 3rd parties, which may be worse than a plain `AttributeError`. What do you think?

Comment (tk0miya) [reply sphinx/builders/linkcheck.py]: I know returing empty values will cause the 3rd party extensions not to work. But it's better than a crash. In other words, my priority is 1) 100% compatible 2) not working 3) crashed.

Comment (tk0miya) [reply sphinx/builders/linkcheck.py]: I found a way to keep compatibility. So I reverted this change.

File: sphinx/builders/linkcheck.py
Comment (francoisfreitag) [inline sphinx/builders/linkcheck.py]:
```suggestion
            "%s.%s is deprecated." % (self.__class__.__name__, "check_thread"),
```

File: sphinx/builders/linkcheck.py
Comment (francoisfreitag) [inline sphinx/builders/linkcheck.py]:
```suggestion
            "%s.%s is deprecated." % (self.__class__.__name__, "limit_rate"),
```

File: sphinx/builders/linkcheck.py
Comment (tk0miya) [inline sphinx/builders/linkcheck.py:254]: Note: I'll separate this from the builder in the next step.

**Comment by tk0miya** [comment]
Now ready to merge this. @francoisfreitag Do you have any comment?

File: sphinx/builders/linkcheck.py
Comment (francoisfreitag) [reply sphinx/builders/linkcheck.py:288]: I think these attributes should be refactored? At least `app` is not used.

File: sphinx/builders/linkcheck.py
Comment (francoisfreitag) [inline sphinx/builders/linkcheck.py]:
```suggestion
    """A worker class for checking the availability of hyperlinks."""
```

File: sphinx/builders/linkcheck.py
Comment (tk0miya) [reply sphinx/builders/linkcheck.py:288]: Sorry for the lack of words. I'll do it on the another PR after merging this. Because it's hard to read the diff until merging this.

**Review by francoisfreitag — APPROVED** [review]
Otherwise LGTM.

**Comment by tk0miya** [comment]
@francoisfreitag Thank you for reviewing. I'll merge this soon after CI passed.

**Comment by tk0miya** [comment]
Merged!

**Pull request merged**
